File: content-solutions/dfs/recipe-loop.js
function solution(n, rules) {
  const graph = Array.from({ length: n }, () => []);
  for (const [a, b] of rules) graph[a].push(b);
  const state = new Array(n).fill(0);
  function hasCycle(v) {
    state[v] = 1;
    for (const w of graph[v]) {
      if (state[w] === 1) return true;
      if (state[w] === 0 && hasCycle(w)) return true;
    }
    state[v] = 2;
    return false;
  }
  for (let v = 0; v < n; v += 1) {
    if (state[v] === 0 && hasCycle(v)) return false;
  }
  return true;
}

File: content-solutions/dfs/recipe-loop.java
import java.util.*;

class Solution {
    private List<List<Integer>> graph;
    private int[] state;

    boolean hasCycle(int v) {
        state[v] = 1;
        for (int w : graph.get(v)) {
            if (state[w] == 1) return true;
            if (state[w] == 0 && hasCycle(w)) return true;
        }
        state[v] = 2;
        return false;
    }

    public boolean solution(int n, int[][] rules) {
        graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        for (int[] e : rules) {
            graph.get(e[0]).add(e[1]);
        }
        state = new int[n];
        for (int v = 0; v < n; v++) {
            if (state[v] == 0 && hasCycle(v)) return false;
        }
        return true;
    }
}
